elastic/kibana
# [Stateful sidenav] Fix breadcrumbs

In this PR I've fixed some of the breadcrumbs for the new solution navigation.

From the linked issue, the goal was mainly to fix the "dashboards" and "spaces" plugin breadcrumbs. I also managed to fix the security solution breadcrumbs.

## How to test

* Add the following in your `kibana.dev.yml`

```yml
xpack.cloud.id: "ftr_fake_cloud_id:aGVsbG8uY29tOjQ0MyRFUzEyM2FiYyRrYm4xMjNhYmM="
xpack.cloud.base_url: "https://cloud.elastic.co"
```
* Launch Kibana and set the "search" solution on the default space
* Navigate to dashboards > dashboard item and make sure the breadcrumbs works as expected
* Navigate to spaces > edit | create space and make sure the breadcrumbs works as expected
* Navigate to the security pages and make sure the breadcrumbs works as expected
* Navigate the stack management pages and make sure the breadcrumbs works as expected

## Notes on implementation

We have 2 chrome style: "classic" and "project".

The "classic" style is the Kibana that we have known for years.
The "project" was initially only for Serverless project, but with the launch of the new solution navigation, it is also used for stateful.

When we are on classic, the API that every plugin calls is `core.chrome.setBreadcrumbs`, and passes **the full** breadcrumbs to render. For serverless we call `serverlessPlugin.setBreadcrumbs` and plugins only call it with the last bit of the breadcrumb to render (as most of the breadcrumbs is already constructed from the navigation tree). This serverless API is only available on serverless so we can't use it in stateful.

I've then extended the `core.chrome.setBreadcrumbs` api and allow a second optional argument to be passed

```ts
core.chrome.setBreadcrumbs([ ... /* classic breadcrumbs */], {
  project: {
    value: [ ... ], // the value of the breadcrumbs when we are in the solution navigation
    absolute: true, // optional flag to indicate if the value is absolute or no (default "false")
  }
})
```

with the following contract:



```ts
interface ChromeSetBreadcrumbsParams {
  /**
   * Declare the breadcrumbs for the project/solution type navigation in stateful.
   * Those breadcrumbs correspond to the serverless breadcrumbs declaration.
   */
  project?: {
    /**
     * The breadcrumb value to set. Can be a single breadcrumb or an array of breadcrumbs.
     */
    value?: ChromeBreadcrumb | ChromeBreadcrumb[];
    /**
     * Indicates whether the breadcrumb should be absolute (replaces the full path) or relative.
     * @default false
     */
    absolute?: boolean;
  };
}
```

Fixes https://github.com/elastic/kibana-team/issues/1147
